File: 207-course-schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        self.graphs = self.build_graph(numCourses, prerequisites)
        
        # define variables
        self.visited = [False] * numCourses
        self.onpath = [False] * numCourses
        self.cycle = False

        for node in range(numCourses):
            self.dfs(node)

        return not self.cycle

    def build_graph(self, numCourses: int, prerequisites: List[List[int]]) -> List[List[int]]:
        graphs = [list() for _ in range(numCourses)]
        for edge in prerequisites:
            _from, _to = edge[1], edge[0]
            graphs[_from].append(_to)
        return graphs

    def dfs(self, src: int) -> None:
        if self.onpath[src]:
            self.cycle = True
        
        if self.visited[src] or self.cycle:
            return
        
        self.visited[src] = True
        self.onpath[src] = True
        
        for n in self.graphs[src]:
            self.dfs(n)
        
        self.onpath[src] = False

File: 207-course-schedule/solution.cpp
class Solution {
public:
    vector<vector<int>> graphs;
    vector<bool> visited, onpath;
    bool cycle;

    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        graphs = build_graph(numCourses, prerequisites);
        
        // define variables
        visited.assign(numCourses, false);
        onpath.assign(numCourses, false);
        cycle = false;

        for (int node = 0; node < numCourses; node++) {
            dfs(node);
        }

        return !cycle;
    }

    vector<vector<int>> build_graph(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graphs(numCourses, vector<int>());
        for (const auto& edge : prerequisites) {
            int from = edge[1], to = edge[0];
            graphs[from].push_back(to);
        }
        return graphs;
    }

    void dfs(int src) {
        if (onpath[src]) {
            cycle = true;
            return;
        }
        if (visited[src] || cycle) {
            return;
        }

        visited[src] = true;
        onpath[src] = true;

        for (const auto& n : graphs[src]) {
            dfs(n);
        }

        onpath[src] = false;
    }
};
